[AST] Add arrays and non-deterministic 'initArray'

wingedtrashcan opened the pull request:
This adds `getAt`, `setAt` and, most importantly, `initArray` operations. The last one is non-deterministic, it should be clear from the comments in the code how it works.

This isn't done completely yet, some polishing and a self-review are required, so perhaps don't review yet, although all the main stuff should be fine and reasonably documented.

allezlom13 commented on src/Nightfall/Targets/Miden.hs:
```diff
@@ -69,14 +84,29 @@ transpile zkProg = do
+                                    (adviceMapExt ctx)
+                                    (_adviceMap inputs)
+                            }
+                        -- TODO: no, we're not supposed to ignore inputs that the program needs just
```
Planning to implement a fix in this PR or later?

wingedtrashcan replied:
This one ideally in a later PR, since it's rather unrelated to what this PR does, I just lumped everything together, because it was easier than creating separate PRs just to realize I missed a piece of functionality etc.

allezlom13 commented on src/Nightfall/Targets/Miden.hs:
```diff
@@ -124,21 +155,119 @@ transpileStatement (AssignVar varname e) = do
+                , variables = Map.insert var arrPos vars
+                }
+
+    -- TODO: should transpilation work in IO instead or can we treat 'runMiden' as "morally-pure"?
```
I think it's fine to have IO there, no need to hide it IMO.

wingedtrashcan replied:
What I'm trying to ask there is whether `transpile` should run in `IO` for `runMiden` not to require `unsafePerformIO` there. Do you think I should change `transpile` to run in `IO`? I'm fine either way, although I'm not a fan of `State` + `IO`, so in that case I'd make the state an `IORef` (or multiple `IORef`s or `MVar`s or whatever).

wingedtrashcan replied:
> Yeah, I'm saying it's fine to have `transpile` run in `IO` yes :)

Got it, OK, it's for a follow-up then.

allezlom13 reviewed (approved):
Great stuffs, LGTM, thanks!

wingedtrashcan commented on src/Nightfall/Lang/Internal/Types.hs:
```diff
@@ -63,13 +68,17 @@ data Statement_ =
     -- | Variable assignment
     | AssignVar VarName Expr_     -- ^ a <- 368
 
+    | SetAt VarName Expr_ Expr_
```
Should it be `AssignAt` or `AssignArray` to be consistent with `AssignVar`? `SetAt` is kinda a stupid name.

wingedtrashcan replied:
I'm gonna ignore this bikeshedding until I have a stronger opinion.

wingedtrashcan commented on src/Nightfall/Lang/Internal/Types.hs:
```diff
@@ -63,13 +68,17 @@ data Statement_ =
     -- | Naked function call
     | NakedCall FunName [Expr_]
 
+    | InitArray VarName [Felt]
```
Needs a Note on the design, especially the CPU vs memory tradeoffs.

wingedtrashcan commented on src/Nightfall/MASM.hs:
```diff
@@ -64,37 +67,37 @@ ppInstr (While body) = do
   indent $ traverse_ ppInstr body
   "end"
 
-ppInstr (LocStore n) = [ "loc_store." ++ show (unMemoryIndex n) ]
-ppInstr (LocLoad n) = [ "loc_load." ++ show (unMemoryIndex n) ]
+ppInstr (LocStore n) = [ "loc_store." ++ show n ]
+ppInstr (LocLoad n) = [ "loc_load." ++ show n ]
```
It's me screwing up git things somehow resulted in those changes being made again. Not sure how and when I cancelled them last time.

wingedtrashcan commented on src/Nightfall/Targets/Miden.hs:
```diff
@@ -43,12 +52,16 @@ defaultConfig = Config
                                              -- access memory, we keep a map to match them
+    , adviceMapExt :: Map Strict.Text [MidenWord]  -- ^ Entries to add to the inputs file
+    , -- Should be a @Set Strict.Text@, really.
+      importExt :: [Lazy.Text]             -- ^ Imports to add to the final program
```
Also the whole situation with `String` vs `Lazy.Text` vs `Strict.Text` should be resolved, currently it's rather messy.

wingedtrashcan replied:
Dropped `Lazy.Text`, so it's only `String` vs `Strict.Text` (vs `DList String`) now.

wingedtrashcan commented on src/Nightfall/Targets/Miden.hs:
```diff
@@ -31,6 +35,11 @@ foldMapA
+consUnlessElem :: Eq a => a -> [a] -> [a]
+consUnlessElem x xs
+    | x `elem` xs = xs
+    | otherwise   = x : xs
```
Remove this nonsense once `importExt` is a `Set`.

wingedtrashcan commented on src/Nightfall/Targets/Miden.hs:
```diff
@@ -124,21 +155,119 @@ transpileStatement (AssignVar varname e) = do
+                , -- Add the position of the array to @i@ to make the top felt on the stack peek at
+                  -- @arr[i]@. Unless the array starts at the very beginning of memory, in which
+                  -- case do nothing.
+                  [MASM.Add . Just $ fromInteger arrPos | arrPos /= 0]
```
That `/= 0` logic should probably go into some combinator, so that we don't create expressions like `x + 0` in any other place too.

wingedtrashcan commented on src/Nightfall/Targets/Miden.hs:
```diff
@@ -124,21 +155,119 @@ transpileStatement (AssignVar varname e) = do
+transpileStatement (SetAt arr i val) = do
+    vars <- gets variables
+    -- TODO: also check that @arr@ is an array.
+    case unMemoryIndex <$> Map.lookup arr vars of
+        Nothing -> error $ "variable \"" ++ arr ++ "\" unknown (undeclared)"
+        Just arrPos -> do
+            shouldTrace <- gets (cfgTraceVariablesUsage . config)
+            let traceVar =
+                    [ MASM.Comment $ "an element of '" <> LText.pack arr <> "' number"
+                    | shouldTrace
+                    ]
```
This should be abstracted into an auxiliary function or something.

wingedtrashcan replied:
I'm gonna postpone this until I understand how to handle tracing best.

wingedtrashcan commented on src/Nightfall/Targets/Miden.hs:
```diff
@@ -124,21 +155,119 @@ transpileStatement (AssignVar varname e) = do
-    return $ args' <> [ MASM.Exec . Text.pack $ fname ]
+    return $ args' <> [ MASM.Exec . LText.pack $ fname ]
+
+transpileStatement (InitArray var inputs) = do
```
Can this one be split into smaller pieces? It's a bit too long. Maybe not worth it.

wingedtrashcan commented on src/Nightfall/Targets/Miden.hs:
```diff
@@ -124,21 +155,119 @@ transpileStatement (AssignVar varname e) = do
+    -- The position of the zeroth element of the array (and therefore the array itself).
+    arrPos <- gets memPos
+
+    -- Allocate memory for all the elements of the array at once.
+    modify $ \ctx ->
+        let vars = variables ctx
+        in if Map.member var vars
+            then error $ "Variable '" ++ var ++ "' has already been declared!"
+            else ctx
+                { memPos = case toMemoryIndex $ unMemoryIndex arrPos + fromIntegral numWords of
+                    Nothing      -> error "Out of Miden's memory"
+                    Just memPos' -> memPos'
+                , variables = Map.insert var arrPos vars
+                }
```
Should use an auxiliary function too.

wingedtrashcan commented on src/Nightfall/Targets/Miden.hs:
```diff
@@ -124,21 +155,119 @@ transpileStatement (AssignVar varname e) = do
+    -- @advice_stack@ and subsequently loaded into memory via @exec.mem::pipe_words_to_memory@,
+    -- which results in the top word on the stack being the hash in reversed form (so we
+    -- additionally reverse it to get the actual hash).
+    hash <- either error pure . fmap (reverse . take 4) . unsafePerformIO . runMiden DontKeep $ do
+        let instrs =
+                [ Push [0, fromIntegral numWords]
+                  -- Stack: [numWords, 0, <zeros>]
+                , Exec "mem::pipe_words_to_memory"
+                  -- Stack: [Hash, memPos', <zeros>]
+                ]
+        Module
+            ["std::mem"]  -- @std::mem@ provides @mem::pipe_words_to_memory@ used above.
+            Map.empty
+            (Program instrs)
+            (Left $ SecretInputs (inputsAsWords >>= midenWordToList) Map.empty)
```
This can clearly be a separate definition.

wingedtrashcan commented on src/Nightfall/Targets/Miden.hs:
```diff
@@ -124,21 +155,119 @@ transpileStatement (AssignVar varname e) = do
+            [ MASM.Comment $ "var " <> LText.pack var
+            | cgfTraceVariablesDecl $ config ctx
+            ]
+    pure $ traceVar ++
```
The comment should clearly be improved.

wingedtrashcan commented on src/Nightfall/Targets/Miden.hs:
```diff
@@ -190,15 +321,41 @@ transpileExpr (VarB varname) = do
+                    | shouldTrace
+                    ]
+            iInstrs <- transpileExpr i
+            return $ concat
+                [ traceVar
+                , iInstrs
+                , -- Add the position of the array to @i@ to make the top felt on the stack peek at
+                  -- @arr[i]@. Unless the array starts at the very beginning of memory, in which
+                  -- case do nothing.
+                  [MASM.Add . Just $ fromInteger arrPos | arrPos /= 0]
+                -- TODO: add (optional) bounds checking.
+                , [MemLoad Nothing]
+                ]
```
Hm... the comment doesn't really specify how many lines are related to what the comment says. That sounds like it can be really confusing. Can we somehow make it visually clear how many lines of Miden code are related to the comment? Using indentation or opening and closing tags or something.

allezlom13 replied:
It's fine like this IMO. When you read the comment you understand what it applies to.

wingedtrashcan replied:
> It's fine like this IMO. When you read the comment you understand what it applies to.

I currently don't, but I haven't really tried properly yet. It's my style of code review though: say out loud whatever I think and then address the stuff that seems worth attention.

wingedtrashcan commented on test/Main.hs:
```diff
@@ -24,14 +27,15 @@ import System.IO.Unsafe
-replaceSecretInputsFileWithContents modul = modul
+-- TODO: handle @advice_map@ too.
+replaceSecretInputsFileWithItsContents :: Module -> Module
+replaceSecretInputsFileWithItsContents modul = modul
```
Play stupid games, win stupid prizes. This should be removed in favor of something sensible.

wingedtrashcan commented on src/Nightfall/MASM/Miden.hs:
```diff
@@ -23,25 +25,44 @@ whenKeep k f = case k of
 
+-- TODO: use @prettyprinter@?
+displaySecretInputs :: SecretInputs -> String
+displaySecretInputs (SecretInputs advStack advMap) = unlines $ concat
```
Add a test.

wingedtrashcan commented:
OK, comments that I wanted to address in this PR are now addressed. The rest are either for a follow-up or some future work.

@allezlom13 please rereview. The algorithmic parts have not changed, I just added docs and tests, plus refactored the code a bit.
